pilot/pkg/networking/core/v1alpha3/listener.go

```diff
@@ -1653,16 +1703,28 @@ type buildListenerOpts struct {
 	needHTTPInspector bool
 }
 
-func buildHTTPConnectionManager(node *model.Proxy, env *model.Environment, httpOpts *httpListenerOpts,
+func buildHTTPConnectionManager(pluginParams *plugin.InputParams, env *model.Environment, httpOpts *httpListenerOpts,
 	httpFilters []*http_conn.HttpFilter) *http_conn.HttpConnectionManager {
-
 	filters := make([]*http_conn.HttpFilter, len(httpFilters))
 	copy(filters, httpFilters)
 
 	if httpOpts.addGRPCWebFilter {
 		filters = append(filters, &http_conn.HttpFilter{Name: wellknown.GRPCWeb})
 	}
 
+	// append ALPN HTTP filter in HTTP connection manager for outbound listener only.
+	if pluginParams.ListenerCategory == networking.EnvoyFilter_SIDECAR_OUTBOUND ||
+		pluginParams.DeprecatedListenerCategory == networking.EnvoyFilter_DeprecatedListenerMatch_SIDECAR_OUTBOUND {
+		filters = append(filters, &http_conn.HttpFilter{
+			Name: AlpnFilterName,
+			ConfigType: &http_conn.HttpFilter_TypedConfig{
+				TypedConfig: util.MessageToAny(&alpn_filter.FilterConfig{
```

Comment:
https://github.com/istio/istio/pull/17470 will change this but technically we should be checking `DisableXDSMarshalingToAny` here.. that PR will probably land anyways though so maybe not relevant

Reply:
Add the guard.